compiler/cbmc/src/cbmc_string.rs

```diff
@@ -0,0 +1,117 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+use lazy_static::lazy_static;
+use std::sync::Mutex;
+use string_interner::symbol::SymbolU32;
+use string_interner::StringInterner;
+
+/// This class implements an interner for Strings.
+/// CBMC objects to have a large number of strings which refer to names: symbols, files, etc.
+/// These tend to be reused many times, which causes signifcant memory usage.
+/// If we intern the strings, each unique string is only allocated once, saving memory.
+/// Since they're referred to by index, InternedStrings become `Copy`, which simplifies APIs.
+/// The downside is that interned strings live the lifetime of the execution.
+/// So you should only intern strings that will be used in long-lived data-structures, not temps.
+///
+/// We use a single global string interner, which is protected by a Mutex (i.e. threadsafe).
```

Comment:
How much does this slow compilation on crates with many dependencies?

Reply:
Actually seems to slightly speed it up!